hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -0,0 +1,212 @@
+					// Skip Enum definitions as we use the base type for storage
+					continue
+				}
+
+				def, err := visitor.VisitDefinition(d, vc)
+				if err != nil {
+					errs = append(errs, err)
+					continue
+				}
+
+				finalDefn := def.WithDescription(descriptionForStorageVariant(d))
+				storageTypes[finalDefn.Name()] = finalDefn
+
+			}
+
+			if len(errs) > 0 {
+				err := kerrors.NewAggregate(errs)
+				return nil, err
+			}
+
+			types.AddTypes(storageTypes)
+
+			return types, nil
+		},
+	}
+}
+
+// makeStorageTypesVisitor returns a TypeVisitor to do the creation of dedicated storage types
+func makeStorageTypesVisitor(types astmodel.Types) astmodel.TypeVisitor {
+	factory := &StorageTypeFactory{
+		types: types,
+	}
+
+	result := astmodel.MakeTypeVisitor()
+	result.VisitTypeName = factory.visitTypeName
+	result.VisitObjectType = factory.visitObjectType
+	result.VisitArmType = factory.visitArmType
+	//result.VisitResourceType = factory.visitResourceType
+
+	return result
+}
+
+type StorageTypeFactory struct {
+	types astmodel.Types
+}
+
+func (factory *StorageTypeFactory) visitTypeName(_ *astmodel.TypeVisitor, name astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+
+	// Resolve the type name to the actual referenced type
+	actualDefinition, actualDefinitionFound := factory.types[name]
+
+	// Check for property specific handling
+	if vc.property != nil && actualDefinitionFound {
+
+		if et, ok := actualDefinition.Type().(*astmodel.EnumType); ok {
+			// Property type refers to an enum, so we use the base type instead
+			return et.BaseType(), nil
+		}
+	}
+
+	// Map the type name into our storage namespace
+	visitedName, err := factory.mapTypeName(name)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to map name into storage namespace")
+	}
+
+	return visitedName, nil
+}
+
+func (factory *StorageTypeFactory) visitObjectType(
+	visitor *astmodel.TypeVisitor,
+	object *astmodel.ObjectType,
+	ctx interface{},
+) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+	oc := vc.forObject(object)
+
+	var errs []error
+	properties := object.Properties()
+	for i, prop := range properties {
+		propertyType, err := visitor.Visit(prop.PropertyType(), oc.forProperty(prop))
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			p := factory.makeStorageProperty(prop, propertyType)
+			properties[i] = p
+		}
+	}
+
+	if len(errs) > 0 {
+		err := kerrors.NewAggregate(errs)
+		return nil, err
+	}
+
+	ot := astmodel.NewObjectType().WithProperties(properties...)
+	return astmodel.MakeStorageType(ot), nil
+}
+
+func (factory *StorageTypeFactory) makeStorageProperty(prop *astmodel.PropertyDefinition, propertyType astmodel.Type) *astmodel.PropertyDefinition {
+	p := prop.WithType(propertyType).
+		MakeOptional().
+		WithoutValidation().
+		WithDescription("")
+
+	// If the property is *string we can simplify it to string
+	// This makes other generated code simpler
```

Comment:
Why is this safe? What about services where there is a difference between "" and `nil`? If we don't preserve that information in the storage type then we'll have a problem, yes?

Reply:
It's safe because we're not storing patches, but the whole thing. With `omitempty` on the types, there's no difference between `""` and **nil** anyway.

Reply:
> With omitempty on the types, there's no difference between "" and nil anyway.

That is not true, [there is a difference](https://play.golang.org/p/qP6nUKTvdXO). I believe we need to preserve this difference as it can have semantic meaning to APIs.

Reply:
I've removed the special casing for strings.